### PR TITLE
Masterbar: Register admin_color as user meta

### LINKED ISSUE #17763: Admin Colors Unification sync process - expose `admin_color` user meta for modification via endpoint
See paYJgx-17t-p2#comment-1391

~Before continuing with this PR see p1605013883088400-slack-jetpack-crew which has important information about where to begin.~

~Create an endpoint for Jetpack (non-WPCom specific) to allow for updating user options.~ Expose the `admin_color` user meta for modification over REST.

See PCYsg-liz-p2

~This should be an endpoint which is proxied to from the public API. I'm doing something similar in D52376-code. We now need to write that new endpoint.~

Should check permissions and capabilities and then update the user option(s) as required.

### PR DESCRIPTION
**Pull request by obenland**
What's better than writing an API endpoint? Not writing an API endpoint.

We can use Core's `/users` endpoint to retrieve and update the `admin_color` user meta, as long as we register it to make it available in the REST API.

Fixes #17763 

#### Changes proposed in this Pull Request:

* Added `class-` prefix to admin-color-scheme class to appease phpcs.
* Registers `admin_color` as a user meta field and makes it available in REST API requests.
* Adds unit tests, covering schema requests, retrieving the color scheme and updating the color scheme.

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

Using [the REST API console](https://wordpress.org/plugins/rest-api-console/):
* In wp-admin, navigate to Tools > REST API console
* Run `GET /wp/v2/users/{id}/?_fields=meta` in the console and make sure the response contains `admin_color`.

##### GET:
image

##### POST
image

To run unit tests:
* Follow [setup guidelines](https://github.com/Automattic/jetpack/blob/master/docs/development-environment.md#unit-testing) if you don't have unit tests set up yet.
* Run `phpunit --testsuite masterbar`.

#### Proposed changelog entry for your changes:

* Adds `admin_color` value to meta information for the users REST API endpoint.

### CODE REVIEW

**Comment by obenland** [comment]
Worth noting that in its current state we'd add that meta field to all sites with the masterbar module active.

/cc @jeherve

**Review by jeherve — CHANGES_REQUESTED** [review]
This tests well for me. Are we okay with the active color scheme becoming publicly available to anyone calling the endpoint, even when not authenticated, or should we limit this to logged in users?

**Comment by getdave** [comment]
@obenland This is really _great_ and I'm able to access via public API proxy with a few tweaks which I'll explain elsewhere...

File: modules/masterbar/admin-color-schemes/class-admin-color-schemes.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	 */
+	public function register_admin_color_meta() {
+		register_meta(
+			'user',
+			'admin_color',
+			array(
+				'default'      => 'fresh',
+				'description'  => 'Slug of the admin color scheme.',
+				'single'       => true,
+				'show_in_rest' => true,
```
[the remark this answers is not in the document]
Comment (getdave) [reply modules/masterbar/admin-color-schemes/class-admin-color-schemes.php]: Good spot fixed.

**Comment by getdave** [comment]
> This tests well for me. Are we okay with the active color scheme becoming publicly available to anyone calling the endpoint, even when not authenticated, or should we limit this to logged in users?

I've added the `auth_callback` for when editing the value https://developer.wordpress.org/reference/functions/register_meta/#parameters

However, I'm not clear on the best way to determine whether or not to expose the field to `GET` / `READ` requests non-auth'd users. Conditionalise the `show_in_rest` based on `current_user_can(`'read')`?

```php
register_meta(
    'user',
    'admin_color',
    array(
        'default'       => 'fresh',
        'description'   => __('Slug of the admin color scheme.', 'jetpack'),
        'single'        => true,
        'show_in_rest'  => current_user_can('read'),
        'type'          => 'string',
        'auth_callback' => function() {
            return current_user_can( 'edit_users' );
        },
    )
);
```

https://wordpress.org/support/article/roles-and-capabilities/#read

File: modules/masterbar/admin-color-schemes/class-admin-color-schemes.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	/**
+	 * Makes admin_color available in users REST API endpoint.
+	 */
+	public function register_admin_color_meta() {
+		register_meta(
+			'user',
+			'admin_color',
+			array(
+				'default'       => 'fresh',
+				'description'   => __('Slug of the admin color scheme.', 'jetpack'),
```
Comment (getdave) [inline modules/masterbar/admin-color-schemes/class-admin-color-schemes.php:31]: Probably needs linting.

**Review by getdave — COMMENTED** [review]
Works really well for me now. The key is to make this work with the public API so we can proxy requests to remote Jetpack sites.

For more on this see D52376-code

Also you'll need to augment the unit tests to cover permissions / capabilities restrictions. I wasn't able to do this as I'm not running Jetpack Docker and didn't want to do the svn dance.

I might setup Jetpack Docker again or ask @glendaviesnz if he knows a way to make Jetpack unit tests work well with `a8c-dev-env`.

File: modules/masterbar/admin-color-schemes/class-admin-color-schemes.php (hunk 2 above)
Comment (obenland) [reply modules/masterbar/admin-color-schemes/class-admin-color-schemes.php:31]: Do you usually skip linting when when you commit?

File: modules/masterbar.php [hunk 3]
```diff
@@ -14,7 +14,7 @@
 namespace Automattic\Jetpack\Dashboard_Customizations;
 
 require dirname( __FILE__ ) . '/masterbar/masterbar/masterbar.php';
```
Comment (jeherve) [inline modules/masterbar.php:16]: Let's fix that one as well.

```suggestion
require __DIR__ . '/masterbar/masterbar/masterbar.php';
```

File: modules/masterbar/admin-color-schemes/class-admin-color-schemes.php (hunk 2 above)
Comment (jeherve) [reply modules/masterbar/admin-color-schemes/class-admin-color-schemes.php:31]: I'd be curious as well. :) You shouldn't have been able to commit.

```suggestion
				'description'   => __( 'Slug of the admin color scheme.', 'jetpack' ),
```

**Review by jeherve — CHANGES_REQUESTED** [review]

File: modules/masterbar/admin-color-schemes/class-admin-color-schemes.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+				'single'        => true,
+				'show_in_rest'  => current_user_can('read'),
```
Comment (jeherve) [inline modules/masterbar/admin-color-schemes/class-admin-color-schemes.php:33]:
```suggestion
				'show_in_rest'  => current_user_can( 'read' ),
```

File: modules/masterbar/admin-color-schemes/class-admin-color-schemes.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+				'type'          => 'string',
+				'auth_callback' => function() {
```
Comment (jeherve) [inline modules/masterbar/admin-color-schemes/class-admin-color-schemes.php:35]:
```suggestion
				'auth_callback' => function () {
```

File: modules/masterbar/admin-color-schemes/class-admin-color-schemes.php (hunk 4 above)
Comment (obenland) [reply modules/masterbar/admin-color-schemes/class-admin-color-schemes.php:33]: It doesn't look like we can limit `show_in_rest` to only be true for "self requests". But we can hide the value. Would that be worth doing?
```suggestion
				'show_in_rest'  => array(
					'prepare_callback' => function( $value, $request ) {
						if ( current_user_can( 'edit_user', $request->get_param( 'id' ) ) ) {
							$value = '';
						}
						return $value;
					},
				),
```

#### Result:
image

Comment (obenland) [reply modules/masterbar/admin-color-schemes/class-admin-color-schemes.php]: I switched it back to `true` to surface the meta key to unauthenticated OPTIONS requests. @jeherve @getdave Any vetos on that or do you think that's fine?

**Review by jeherve — APPROVED** [review]

**Comment by obenland** [comment]
I had to resolve merge conflicts after #17694. @jeherve @kraftbj Could I get another ✅  on here?

**Review by jeherve — APPROVED** [review]

**Comment by getdave** [comment]
Thanks for everything on this @obenland 🥳

**Comment by obenland** [comment]
r217804-wpcom
r217806-wpcom

**Pull request merged**